### PR TITLE
Introduce a Build class

### PR DESCRIPTION
**Pull request by tizzo**
The code in the container manager at the `runbuild` endpoint is more complicated than it needs to be and the API calls necessary to instantiate and run a task properly have gotten out of control, spread across the `Container` model, and `Task` model.  I believe this is partly due to a missing and helpful `Build` domain model which provides an abstraction for the process of running a series of tasks on a given container. A container has one build (generally, it will be possible to run two builds on one container but probably not very useful), and a build has one or more tasks.

This is an ongoing work in progress but will simplify the completion of #10.

### CODE REVIEW

File: lib/Build.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
+var events = require('events')
```
Comment (zanchin) [inline lib/Build.js]: We should add `"use strict";` for all new files so we can use `let` and other ES6 goodness.

Comment (tizzo) [inline lib/Build.js]: Yep, as per the new coding standards it's requried :-).

**Comment by tortillaj** [comment]
Out of curiosity, what happens if a user's auth token isn't valid?

**Comment by tizzo** [comment]
Yep, the build would fail because the git downloader wouldn't be able to do it's work. Then likely all other steps would fail too.

File: lib/Build.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    // The array of provision steps that must be run in this build.
+    this.steps = [];
+    // The step currently being run when iterating on steps.
+    this.position = 0;
+    this.stream = through2.obj();
+    this.endedStreams = [];
+    this.stepErrors = [];
+    // Used when iterating over the steps to determine the current step that we're running.
+    this.currentStepIndex = 0;
+    events.EventEmitter.call(this);
```
Comment (zanchin) [inline lib/Build.js]: Is this still necessary with the `super()` call above?

File: lib/Build.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    this.container = null;
+    // The array of provision steps that must be run in this build.
+    this.steps = [];
+    // The step currently being run when iterating on steps.
+    this.position = 0;
+    this.stream = through2.obj();
+    this.endedStreams = [];
+    this.stepErrors = [];
+    // Used when iterating over the steps to determine the current step that we're running.
+    this.currentStepIndex = 0;
```
Comment (zanchin) [inline lib/Build.js]: Is `currentStepIndex` the same thing as `this.position` a few lines above?

File: lib/plugins/TaskRunner/AbstractPlugin.js [hunk 4]
```diff
@@ -1,220 +1,302 @@
...
+      logContext.execId = this.exec.id;
+    }
+    if (this.build) {
+      logContext.buildId = this.build.id;
+    }
+    return logContext;
+  }
 
-      return errorHandler(error, `${self.name}: Failed to exec command: ${error.message}`);
+  get ellapsedTime() {
```
Comment (zanchin) [inline lib/plugins/TaskRunner/AbstractPlugin.js]: typo: `ellapsedTime` -> `elapsedTime`

File: lib/plugins/TaskRunner/AbstractPlugin.js [hunk 5]
```diff
@@ -1,220 +1,302 @@
...
-  var ret = {};
-  var props = this._json_attrs || ['id', 'name', 'plugin', 'timeout', 'options', 'result'];
-  for (var i = 0; i < props.length; i++) {
-    ret[props[i]] = this[props[i]];
-  }
-  return ret;
-};
 
+    // This will be the reference to the timer returned by setTimeout().
+    this[_timer] = null;
```
Comment (zanchin) [inline lib/plugins/TaskRunner/AbstractPlugin.js]: Should this be `this['_timer'] = null;` instead?

**Pull request closed without merging**